### PR TITLE
parallel build

### PR DESCRIPTION
**Pull request by szagoruyko**

### CODE REVIEW

**Comment by zhangxiangxiao** [comment]
I wish there is a 'like' button on github! We had something like this in an internal cunn at NYU a long time ago, but never created a pull request. This makes writing new modules and debugging much much faster!

File: PReLU.cu [hunk 1]
```diff
@@ -2,12 +2,22 @@
...
       i < (n);                                       \
       i += blockDim.x * gridDim.x)
 
+// Use 1024 threads per block, which requires cuda sm_2x or above
+const int CUDA_NUM_THREADS = 1024;
+
+// CUDA: number of blocks for threads.
+inline int GET_BLOCKS(const int N) {
+  return (N + CUDA_NUM_THREADS - 1) / CUDA_NUM_THREADS;
+}
```
Comment (dominikgrewe) [inline PReLU.cu]: Instead of duplicating this wherever it's used, it'd be better to put this into a shared header file.

File: init.cu [hunk 2]
```diff
@@ -10,37 +10,6 @@
 
 #include "utils.c"
```
Comment (dominikgrewe) [inline init.cu]: While we're at it, "utils.c" should also be compiled separately.

**Comment by dominikgrewe** [comment]
Thanks a lot! This was long overdue :)

Do we really need to make every function global though? Isn't it enough to make the `cunn_*_init` functions global?

**Comment by szagoruyko** [comment]
@zhangxiangxiao agree!
@dominikgrewe you're right, that was an overkill. Will update now.

File: PReLU.cu (hunk 1 above)
Comment (szagoruyko) [inline PReLU.cu]: do you have a name for it? :)

Comment (dominikgrewe) [inline PReLU.cu]: common.h ?

Comment (szagoruyko) [inline PReLU.cu]: I was thinking about something fancy. Do you think CUDA_NUM_THREADS should be there too?

Comment (dominikgrewe) [inline PReLU.cu]: Yes, I think so. In fact, with the current definition of `GET_BLOCKS` it has to be there.

**Comment by hughperkins** [comment]
cool :-)

**Comment by szagoruyko** [comment]
@dominikgrewe updated

**Comment by dominikgrewe** [comment]
Thanks! Is there a reason why you renamed `utils.c` to `utils.cu`? There's just plain C code in there, so calling it `utils.c` makes more sense to me.

**Comment by szagoruyko** [comment]
because ```extern "C"``` has to used to include "lua.h" in .cu files and caused errors in utils.c.
Actually init.cu could be renamed to init.c too, but there are build errors if I do that.

**Comment by hughperkins** [comment]
```
ifdef __cplusplus
    extern"C" {
#endif
```
 ?

**Comment by szagoruyko** [comment]
@hughperkins thanks Hugh, but then header file has c linkage and all the cu files have c++ linkage and we'd have to add extern "C" to all cunn_*_init files.
I think we should leave them in with cu extension.

**Comment by dominikgrewe** [comment]
Sounds good to me. We already have init.cu, so it's at least consistent.

If you squash your commit this is ready for merge.

**Comment by szagoruyko** [comment]
done

**Comment by dominikgrewe** [comment]
Thanks!

**Pull request merged**